src/pip/_internal/utils/filesystem.py

```diff
@@ -169,3 +170,67 @@ def _test_writable_dir_win(path):
     raise EnvironmentError(
         'Unexpected condition testing for writable directory'
     )
+
+
+def find_files(path, pattern):
+    # type: (str, str) -> List[str]
+    """Returns a list of absolute paths of files beneath path, recursively,
+    with filenames which match the UNIX-style shell glob pattern."""
+    result = []  # type: List[str]
+    for root, dirs, files in os.walk(path):
+        matches = fnmatch.filter(files, pattern)
+        result.extend(os.path.join(root, f) for f in matches)
+    return result
+
+
+def _friendly_size(size):
+    # type: (Union[float, int]) -> str
+
+    # Explicitly convert `size` to a float, for consistent behavior
+    # between Python 2 and Python 3.
+    size = float(size)
+
+    suffix = 'B'
+    if size > 1000:
+        size /= 1000
+        suffix = 'KB'
+
+    if size > 1000:
+        size /= 1000
+        suffix = 'MB'
+
+    if size > 1000:
+        size /= 1000
+        suffix = 'GB'
+
+    size = round(size, 1)
+
+    return '{} {}'.format(size, suffix)
+
+
+def file_size(path):
+    # type: (str) -> Union[int, float]
+    # If it's a symlink, return 0.
+    if os.path.islink(path):
+        return 0
```

Comment:
Arguably... symlinks do have a size -- the number of characters in the reference. :P

Reply:
I was going off what `du` reports on my (Linux) system, but apparently different tools &mdash; and even different invocations of the same tool &mdash; report different things on my system!

E.g., in this case I just tried, `du` reports 0, `ls -s` reports 0, but `ls -l` reports 42 (the length of the reference):

```
~$ file ~/.config/nvim/init.vim
/home/puppy/.config/nvim/init.vim: symbolic link to /home/puppy/dotfiles/.config/nvim/init.vim
~$ du ~/.config/nvim/init.vim
0       /home/puppy/.config/nvim/init.vim
~$ ls -s ~/.config/nvim/init.vim
0 /home/puppy/.config/nvim/init.vim
~$ ls -l ~/.config/nvim/init.vim
lrwxrwxrwx. 1 puppy puppy 42 Jan 21 08:20 /home/puppy/.config/nvim/init.vim -> /home/puppy/dotfiles/.config/nvim/init.vim
~$
```

I _think_ what's happening is that, excluding `ls -l`, all the things I tried were rounding to the nearest KB.

I tried it on FreeBSD and got completely different results. On there, `du` reports 1(?!), `ls -s`  reports the length of the reference, and `ls -l` also reports the length of the reference.

So I'm not really sure what's most appropriate here. The two operating systems I have easy access to return different results, `os.path.getsize` resolves it (meaning it may count things multiple times), and I don't know what other operating systems do. :upside_down_face:

Reply:
I think it's probably irrelevant - this is only used in `pip cache` at the moment, which (as far as I know) won't contain symlinks anyway.

I'd suggest leaving it as it is. If someone comes along with a use case for this function that needs a different answer for symlinks, let's revisit it then.

Reply:
Sounds good to me. Marking this thread as resolved. :+1: